GitHub Actions workflow to publish specification to GitHub Pages

joshuagl opened the pull request:
Here's a simple collection of scripts and GitHub Actions workflows to build and publish the bikeshed specification when it is updated.

On pushes to master the workflow will:
* create a release
* build that release
* publish it to both a versioned subdirectory of the gh-pages branch _and_ the latest subdirectory of the gh-pages branch
* build a simple HTML index to point to all published versions of the bikshed specification (and the draft version)

On pushes to draft the workflow will:
* build the draft
* publish it to a draft subdirectory of the gh-pages branch

You can see an example of the automation here in action at https://joshuagl.github.io/specification/

**Note**: we need GitHub Actions and GitHub Pages enabling for the repository for this to work.

Fixes #145

joshuagl commented:
You can see an example of this GHA workflow in action on my personal fork here https://github.com/joshuagl/specification/pull/5/checks?check_run_id=1870772911

joshuagl commented:
Also an example of a failed run, because date was not changed, here: https://github.com/joshuagl/specification/pull/6/checks?check_run_id=1903153676

joshuagl commented:
I've updated this PR to not only convert to GitHub Actions, but also to include workflows to build, release (in the case of master) & publish the specification on pushes to the master and draft branches.

Please take a look @lukpueh, @trishankatdatadog and @mnm678.

lukpueh commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,18 @@
+
+jobs:
+  main:
+    name: Build, Validate and Deploy
```
We only run the release checker here, right? Is this name accurate?

joshuagl replied:
Nice catch, I will update the name.

lukpueh commented on .github/workflows/pr.yml:
```diff
@@ -0,0 +1,18 @@
+name: CI
+on:
+  pull_request: {}
```
f9fd90af5ad8233ccab54baf8a61a35e9134d311 says "we will configure CI to only run check_release on PRs to master". Do we have to configure this here?

joshuagl replied:
We do! I will fix that, thank you.

lukpueh reviewed (commented):
Very impressive work, @joshuagl, and as usual, very nicely laid out (re commit separation and messages)!! 🎉 

I'd like to take another look at the release and draft workflow files with a fresh set of eyes on Monday, but on a quick glance they seem to do what they should.

joshuagl commented on build_index.py:
```diff
@@ -0,0 +1,203 @@
+        sys.exit()
+
+def build_index():
+    # sanity_check()
```
Good question! I think it does not make sense to enforce the sanity check. Our GHA workflow ensures we are on the correct branch, and we should let folks using the script offline decide where to run it. Indeed, I had to comment this out for testing and completely forgot I had even written it...

joshuagl commented on get_version.py:
```diff
@@ -0,0 +1,44 @@
+<Purpose>
+  Quick and dirty script to get the version number from tuf-spec.md
+
+  Unfortunately GNU grep and BSD grep take different options...
```
That's right, I wanted to ensure we could replicate the GHA workflow outside of on an Actions runner with minimal effort.

joshuagl commented:
Thank you for the detailed review @lukpueh. I've addressed your comments by removing the sanity check code, renaming the PR workflow name and adding a branch specifier to the PR workflow.

lukpueh commented on check_release.py:
```diff
@@ -160,4 +145,4 @@ def main():
 
```
> don't forget to tag the release ..

Not that this ever really reminded someone, but it can definitely go now. :)

joshuagl replied:
Removed in 47869c0

lukpueh reviewed (approved):
LGTM modulo some minor nits that you may address or not. Many thanks for the excellent PR, @joshuagl!

joshuagl commented on .github/workflows/draft.yml:
```diff
@@ -0,0 +1,53 @@
+    - name: Push generated specification
+      run: |
+        git config user.name "TUF Specification Automation"
+        git config user.email tuf-spec-bot@example.com
```
I like the idea of using the mailing list!

joshuagl commented on .github/workflows/draft.yml:
```diff
@@ -0,0 +1,53 @@
+on:
+  workflow_dispatch:
+  push:
+    branches: draft
```
That's right, a merge pushes changes to the branch and results in a 'push' event.

joshuagl commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,81 @@
+        rmdir build
+        make index
+        git add .
+        git commit -m "Publish latest specification"
```
Yes, that would be cool. Not much work at all, I've added it in 948022f

joshuagl commented:
> LGTM modulo some minor nits that you may address or not. Many thanks for the excellent PR, @joshuagl!

Some good nits, I've addressed each in some follow-on commits. Please review the extra 3 patches and re-approve the PR at your convenience.

lukpueh reviewed (approved):
💯

joshuagl commented:
If anyone has the time, I would appreciate a second review on this so we can get it merged and start publishing the newly formatted specification.

mnm678 reviewed (approved):
Looks great, thanks for all the work on this!
